hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+void BleObject::BleEventDispatcher::onGenericEventCallback(on_ble_evt_cb_t cb, void* context) {
+    BleGenericEventHandler evtHandler;
+    evtHandler.handler = cb;
+    evtHandler.context = context;
+    genericEventHandlers_.append(evtHandler);
+}
+
+os_thread_return_t BleObject::BleEventDispatcher::bleEventDispatch(void* param) {
+    BleEventDispatcher* dispatcher = static_cast<BleEventDispatcher*>(param);
+    while (1) {
+        EventMessage msg;
+        if (!os_queue_take(dispatcher->evtQueue_, &msg, CONCURRENT_WAIT_FOREVER, nullptr)) {
+            if (msg.handler) {
+                // Invoke the handler registered by specific application.
+                if (msg.evt.type == BLE_EVT_SCAN_RESULT) {
+                    msg.handler(&msg.evt.params.scan_result, msg.context);
+                } else if (msg.evt.type == BLE_EVT_SVC_DISCOVERED) {
+                    msg.handler(&msg.evt.params.svc_disc, msg.context);
+                } else if (msg.evt.type == BLE_EVT_CHAR_DISCOVERED) {
+                    msg.handler(&msg.evt.params.char_disc, msg.context);
+                }
+            } else {
+                // Just dispatch the event to the application those have subscribed the generic events.
+                for (int i = 0; i < dispatcher->genericEventHandlers_.size(); i++) {
+                    BleGenericEventHandler& evtHandler = dispatcher->genericEventHandlers_[i];
+                    if (evtHandler.handler) {
+                        evtHandler.handler(&msg.evt, evtHandler.context);
+                    }
+                }
+            }
+            // Free the memory if needed.
+            if (msg.hook) {
+                msg.hook(&msg.evt, msg.hookContext);
+            }
         } else {
-            // Pairing is not supported
-            ret = sd_ble_gap_sec_params_reply(conn, BLE_GAP_SEC_STATUS_PAIRING_NOT_SUPP, nullptr, nullptr);
-        }
-        if (ret != NRF_SUCCESS) {
-            LOG(ERROR, "sd_ble_gap_sec_params_reply() failed: %u", (unsigned)ret);
+            LOG(ERROR, "BLE event thread exited.");
+            break;
         }
-        break;
     }
-    case BLE_GAP_EVT_AUTH_STATUS: {
-        const auto status = event->evt.gap_evt.params.auth_status.auth_status;
-        if (status == BLE_GAP_SEC_STATUS_SUCCESS) {
-            LOG_DEBUG(TRACE, "Authentication succeeded");
-        } else {
-            LOG(WARN, "Authentication failed, status: %d", (int)status);
-        }
-        break;
+    os_thread_exit(dispatcher->evtThread_);
+}
+
+typedef struct {
+    BleObject::BleGap* instance;
+} BleGapImpl;
+static BleGapImpl bleGapImpl;
+
+BleObject::BleGap::BleGap(BleObject* obj) : bleObj_(obj) {
+}
+
+int BleObject::BleGap::init(void) {
+    // Set the default device name
+    char devName[32] = {};
+    get_device_name(devName, sizeof(devName));
+    setDeviceName((const uint8_t*)devName, strlen(devName));
+    bleGapImpl.instance = this;
```

Comment:
Why do we need this object? Can't we simply pass `this`? And even that is probably not necessary because `BleObject::instance()->xxx` can be used at all times.

Reply:
This is somehow the section conflict between c and c++ code:
```
src/nRF52840/ble_hal.cpp:2166:26: error: bleGattClient causes a section type conflict with bleGap
     NRF_SDH_BLE_OBSERVER(bleGattClient, 1, processGattClientEvents, &gattcImpl);
                          ^
../third_party/nrf5_sdk/nrf5_sdk/components/libraries/experimental_section_vars/nrf_section.h:149:5: note: in definition of macro 'NRF_SECTION_ITEM_REGISTER'
     section_var __attribute__ ((section("." STRINGIFY(section_name)))) __attribute__((used))
     ^
../third_party/nrf5_sdk/nrf5_sdk/components/softdevice/common/nrf_sdh_ble.h:85:1: note: in expansion of macro 'NRF_SECTION_SET_ITEM_REGISTER'
 NRF_SECTION_SET_ITEM_REGISTER(sdh_ble_observers, _prio, static nrf_sdh_ble_evt_observer_t _name) =  \
 ^
src/nRF52840/ble_hal.cpp:2166:5: note: in expansion of macro 'NRF_SDH_BLE_OBSERVER'
     NRF_SDH_BLE_OBSERVER(bleGattClient, 1, processGattClientEvents, &gattcImpl);
     ^
src/nRF52840/ble_hal.cpp:639:26: note: 'bleGap' was declared here
     NRF_SDH_BLE_OBSERVER(bleGap, 1, processBleGapEvents, this);
                          ^
../third_party/nrf5_sdk/nrf5_sdk/components/libraries/experimental_section_vars/nrf_section.h:149:5: note: in definition of macro 'NRF_SECTION_ITEM_REGISTER'
     section_var __attribute__ ((section("." STRINGIFY(section_name)))) __attribute__((used))
     ^
../third_party/nrf5_sdk/nrf5_sdk/components/softdevice/common/nrf_sdh_ble.h:85:1: note: in expansion of macro 'NRF_SECTION_SET_ITEM_REGISTER'
 NRF_SECTION_SET_ITEM_REGISTER(sdh_ble_observers, _prio, static nrf_sdh_ble_evt_observer_t _name) =  \
 ^
src/nRF52840/ble_hal.cpp:639:5: note: in expansion of macro 'NRF_SDH_BLE_OBSERVER'
     NRF_SDH_BLE_OBSERVER(bleGap, 1, processBleGapEvents, this);
```

Reply:
AFAICR, you can't use `this` with the Nordic's `NRF_SDH_BLE_OBSERVER()` macro. Although a similar code was compiling fine in my tests, at runtime `this` was always set to NULL.

Reply:
Can this be marked as resolved?